tech/post/2017-09-25-Kubernetes-Scheduler.md

```diff
@@ -0,0 +1,145 @@
+- First Scheduler creates ```configFactory``` and then creates ```config```.
+- If it specifies policy-config-file, Scheduler will create ```config``` by policy-config-file(CreateFromConfig),
+- If not, Scheduler will create ```config``` by Scheduler algorithm providers(CreateFromProvider),
+- After Scheduler have got all the parameters, the scheduling work will be executed in ```plugin/pkg/scheduler/scheduler.go```.
+- Scheduler invokes ```scheduleOne``` function and ```getNextPod``` from pod queue.
+- Then Scheduler execute ```schedule``` function, and actually the schedule process will be executed in ```plugin/pkg/scheduler/core/generic_scheduler.go```.
+- If the operation is successful, the Scheduler will get a most fit node for this pod.
+- Then Scheduler will tell the cache to assume that a pod now is running on a given node, even though it hasn't been bound yet. This allows Scheduler to keep scheduling without waiting on binding to occur. 
+- At last Scheduler will bind the pod to its node asynchronously by invoking Kubernetes APIServer Bind interface.
+
+![](../images/2017-09-25-Kubernetes-Scheduler/SchedulerSequence.png)
+
+The Kubernetes Scheduler is able to find the most fit node for pods. The actual scheduling steps include: ```predicates``` and ```priorities```. The default algorithm provider is ```DefaultProvider``` which can be specified by setting the parameter ```--algorithm-provider string``` when starting the Scheduler. 
+
+![](../images/2017-09-25-Kubernetes-Scheduler/AlgorithmProviderSequence.png)
+
+```DefaultProvider``` is registered by two functions ```defaultPredicates``` and ```defaultPriorities```. Also these two functions can be configured by setting the parameter ```--policy-config-file string```. Here is the example about scheduler-policy-config.json.
+
+```
+{
+	"kind" : "Policy",
+	"apiVersion" : "v1",
+	"predicates" : [
+		{"name" : "PodFitsHostPorts"},
+		{"name" : "PodFitsResources"},
+		{"name" : "NoDiskConflict"},
+		{"name" : "NoVolumeZoneConflict"},
+		{"name" : "MatchNodeSelector"},
+		{"name" : "HostName"}
+		],
+	"priorities" : [
+		{"name" : "LeastRequestedPriority", "weight" : 1},
+		{"name" : "BalancedResourceAllocation", "weight" : 1},
+		{"name" : "ServiceSpreadingPriority", "weight" : 1},
+		{"name" : "EqualPriority", "weight" : 1}
+		],
+	"hardPodAffinitySymmetricWeight" : 10
+}
+```
+
+## 4. Scheduler Algorithm
+
+### 4.1 Default Predicates
+
+#### 4.1.1 NoVolumeZoneConflict
+
+Evaluate if the volumes a pod requests are available on the node, given the Zone restrictions.
+
+#### 4.1.2 MaxEBSVolumeCount
+Ensure that the number of attached ElasticBlockStore volumes does not exceed a maximum value (by default, 39, since Amazon recommends a maximum of 40 with one of those 40 reserved for the root volume -- see Amazon's documentation). The maximum value can be controlled by setting the KUBE_MAX_PD_VOLS environment variable.
+
+#### 4.1.3 MaxGCEPDVolumeCount

```

Comment:
Is there a environment variable to limit the volume count from the OpenStack?

Reply:
Not yet